File: Domain/Song.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace Domain;

public class Song
{
    [ForeignKey("AppUser")] public string AppUserId { get; set; } = null!;
    public AppUser AppUser { get; set; } = null!;

    // main song details
    [Key] public int Id { get; set; }
    [Column(TypeName = "varchar(50)")]
    public string SongName { get; set; } = null!;
    public double MasterVolume { get; set; }
    public double DrumVolume { get; set; }
    public double BassVolume { get; set; }
    public double MelodicVolume { get; set; }
    public int Bpm { get; set; }

    // effects
    public Distortion Distortion { get; set; } = null!;
    public Reverb Reverb { get; set; } = null!;
    public Delay Delay { get; set; } = null!;

    // synths
    public BassSynth BassSynth { get; set; } = null!;
    public MelodicSynth MelodicSynth { get; set; } = null!;

    // patterns
    public MelodicPattern MelodicPattern { get; set; } = null!;
    public KitPattern KitPattern { get; set; } = null!;
}

File: client/src/models/song.ts
import { PatternName } from 'tone/build/esm/event/PatternGenerator';
import { Sampler } from 'tone';
import { SimpleWaveType, WaveType } from './types/waveTypes.ts';
import { NoteType } from './types/noteType.ts';
import { FilterType } from './types/filterType.ts';
import { MusicalKey } from './types/musicalKey.ts';
import { MusicalScale } from './types/musicalScale.ts';

export interface Song {
  id?: number;
  songName?: string;
  masterVolume: number;
  drumVolume: number;
  bassVolume: number;
  melodicVolume: number;
  bpm: number;
  distortion: Distortion;
  reverb: Reverb;
  delay: Delay;
  bassSynth: BassSynth;
  melodicSynth: MelodicSynth;
  melodicPattern: MelodicPattern;
  kitPattern: KitPattern;
}

export interface Distortion {
  id?: number;
  presetName?: string;
  mix: number;
  amount: number;
  filterFrequency: number;
  songId?: number;
}

export interface Reverb {
  id?: number;
  presetName?: string;
  mix: number;
  decay: number;
  preDelay: number;
  songId?: number;
}

export interface Delay {
  id?: number;
  presetName?: string;
  mix: number;
  time: number;
  feedback: number;
  songId?: number;
}

export interface BassSynth {
  id?: number;
  presetName?: string;
  attack: number;
  decay: number;
  sustain: number;
  release: number;
  waveform: WaveType;
  filterFrequency: number;
  songId?: number;
}

export interface MelodicSynth {
  id?: number;
  presetName?: string;
  waveform: WaveType;
  attack: number;
  decay: number;
  sustain: number;
  release: number;
  filterFrequency: number;
  filterMod: number;
  filterType: FilterType;
  metal: number;
  chorus: number;
  lfoFrequency: NoteType;
  lfoShape: SimpleWaveType;
  songId?: number;
}

export interface MelodicPattern {
  id?: number;
  presetName?: string;
  key: MusicalKey;
  scale: MusicalScale;
  sequence: number[];
  patternType: PatternName;
  transpose: number;
  songId?: number;
  timeInterval: NoteType;
  noteDuration: NoteType;
  length: number;
}

export interface KitPattern {
  id?: number;
  presetName?: string;
  patternLength: number;
  bdSteps: boolean[];
  sdSteps: boolean[];
  clSteps: boolean[];
  chSteps: boolean[];
  songId?: number;
}

export interface Drum {
  name: DrumNames;
  drumTrackSteps: DrumTrackSteps;
  id: number;
  sampler: Sampler;
}

export type DrumNames = 'BD' | 'SD' | 'CL' | 'CH';

export type DrumTrackSteps = 'bdSteps' | 'sdSteps' | 'clSteps' | 'chSteps';
